app/com/linkedin/drelephant/spark/heuristics/StagesAnalyzer.scala

```diff
@@ -0,0 +1,484 @@
+  * @param data Spark application data
+  */
+private[heuristics] class StagesAnalyzer(
+    private val heuristicConfigurationData: HeuristicConfigurationData,
+    private val data: SparkApplicationData) {
+
+  import ConfigurationUtils._
+
+  // serverity thresholds for execution memory spill
+  private val executionMemorySpillThresholds: SeverityThresholds =
+    SeverityThresholds.parse(heuristicConfigurationData.getParamMap
+      .get(SPARK_STAGE_EXECUTION_MEMORY_SPILL_THRESHOLD_KEY), ascending = true)
+      .getOrElse(DEFAULT_EXECUTION_MEMORY_SPILL_THRESHOLDS)
+
+  // severity thresholds for task skew
+  private val taskSkewThresholds: SeverityThresholds =
+    SeverityThresholds.parse(heuristicConfigurationData.getParamMap
+      .get(SPARK_STAGE_TASK_SKEW_THRESHOLD_KEY), ascending = true)
+      .getOrElse(DEFAULT_TASK_SKEW_THRESHOLDS)
+
+  // severity thresholds for task duration (long running tasks)
+  private val taskDurationThresholds: SeverityThresholds =
+    SeverityThresholds.parse(heuristicConfigurationData.getParamMap
+      .get(SPARK_STAGE_TASK_DURATION_THRESHOLD_KEY), ascending = true)
+      .getOrElse(DEFAULT_TASK_DURATION_THRESHOLDS)
+
+  // severity thresholds for task failures
+  private val taskFailureRateSeverityThresholds: SeverityThresholds =
+    SeverityThresholds.parse(heuristicConfigurationData.getParamMap
+      .get(TASK_FAILURE_RATE_SEVERITY_THRESHOLDS_KEY), ascending = true)
+      .getOrElse(DEFAULT_TASK_FAILURE_RATE_SEVERITY_THRESHOLDS)
+
+  // execution memory spill: threshold for processed data, above which some spill is expected
+  private val maxDataProcessedThreshold = MemoryFormatUtils.stringToBytes(
+    heuristicConfigurationData.getParamMap
+    .getOrDefault(MAX_DATA_PROCESSED_THRESHOLD_KEY, DEFAULT_MAX_DATA_PROCESSED_THRESHOLD))
+
+  // threshold for ratio of max task duration to stage duration, for flagging task skew
+  private val longTaskToStageDurationRatio = heuristicConfigurationData.getParamMap
+    .getOrDefault(LONG_TASK_TO_STAGE_DURATION_RATIO_KEY, DEFAULT_LONG_TASK_TO_STAGE_DURATION_RATIO).toDouble
+
+  // min threshold for median task duration, for flagging task skew
+  private val taskDurationMinThreshold = heuristicConfigurationData.getParamMap
+    .getOrDefault(TASK_SKEW_TASK_DURATION_MIN_THRESHOLD_KEY, DEFAULT_TASK_SKEW_TASK_DURATION_MIN_THRESHOLD).toLong
+
+  // the maximum number of recommended partitions
+  private val maxRecommendedPartitions = heuristicConfigurationData.getParamMap
+    .getOrDefault(MAX_RECOMMENDED_PARTITIONS_KEY, DEFAULT_MAX_RECOMMENDED_PARTITIONS).toInt
+
+  /**
+    * Get the analysis for each stage of the application.
+    *
+    * @param curNumPartitions the configured number of partitions for the application
+    *                         (value of spark.sql.shuffle.partitions).
+    * @return list of analysis results of stages.
+    */
+  def getStageAnalysis(curNumPartitions: Int): Seq[StageAnalysis] = {
```

Comment:
Can fetch the curNumPartitions in the method itself from the `data` (SparkApplicationData)